### ISSUE TITLE
using different atlas

### ISSUE THREAD

**Issue by zswgzx** [Deep-MI/FastSurfer#51]
Hi this tool is very helpful for segmentation and I wonder if it allows for segmenting with a different FS atlas, e.g. a2009s? Haven't seen a similar post but please point out if any. Thanks!

**Comment by zswgzx** [comment]
Some other messages:
```
eval.py:376: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  aseg_lh = gaussian_filter(1000 * np.asarray(pred_prob == 2, dtype=np.float), sigma=3)
eval.py:377: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  aseg_rh = gaussian_filter(1000 * np.asarray(pred_prob == 41, dtype=np.float), sigma=3)
```

**Comment by zswgzx** [comment]
Thank you @LeHenschel . Just curious if DKTatlas would be the only version for output with adding -fsaparc as considering other default FS atlases? So for example, would a2009s labels  be included also for the updated output with the -fsaparc option?

Btw, does fast surfer output the full folder structures for segmentation only, e.g. label or stats, as generated by Freesurfer?

Thanks again! Shengwei

**Comment by m-reuter** [comment]
Hi, with -fsaparc you should get all the FS atlas versions. Also FastSurfer does output full folder structure. It is missing only a few intermediate files which are not required for our processing steps.

**Comment by zswgzx** [comment]
Thank you professor @m-reuter for the update. Would you mind providing more detailed usage as such? Below was the command tried:

cd $fastsurfer
./run_fastsurfer.sh --t1 $t1.nii.gz --sid $id --sd $dest --py python --seg_only --parallel --fsaparc

I only got $dest/{mri,scripts} and only aparc.DKTatlas+aseg.deep.mgz & deep-seg.log. Ideally a full FS folder structure with  $dest/{label, stats, ...} folders would be expected, and aparc.a2009s+aseg.deep.mgz would be expected in folder $dest/mri. Are they achievable and how? Thanks again!

**Comment by LeHenschel** [comment]
Hey Shengwei,

the fsaparc flag only takes effect if you run the entire fastsurfer pipeline (including recon-surf). The --seg_only flag only runs the DL prediction which is limited to the DKTatlas.

So just omit the --seg_only flag and you will get the FreeSurfer conform folder structure including the a2009s atlas:

```bash
cd $fastsurfer
./run_fastsurfer.sh --t1 $t1.nii.gz --sid $id --sd $dest --py python --parallel --fsaparc
```

Hope this helps.

Best,
Leonie

**Comment by zswgzx** [comment]
Got it Leonie, that's very helpful. Let me try and I'll close if I don't have any other issue. Nice work!

**Comment by zswgzx** [comment]
One more question as I'm running it on a Azure VM (6 cpus w/ 56G RAM): does the --thread option only affect the surface pipeline? Setting up the OMP/ITK threads could accelerate the processing time if I understand correctly. But how should I choose? My final goal is to dockerize it for a pre-selected VM size, so I can scale it to a larger population.

**Comment by m-reuter** [comment]
Note that the full pipeline is of course much longer than the 1 minute (GPU) for segmentation. If you run purely CPU also the segmentation will be much slower (like around 30 mins or longer). 
About the --thread option, it only affects the recon_surf pipeline, so basically all FreeSurfer executables that can do parallel processing. Not sure which these are but with 6 CPUs you should definitely try to run in parallel, maybe try 6 and you ran run both hemispheres in parallel (--parallel flag).

**Comment by zswgzx** [comment]
I see, will try using as many threads as the number of cpus. And I'm trying it on a GPU VM so it does run FAST! Another question, hopefully the last, that comes to mind is that should I re-run if there's any manual correction(s) made? FS has some flags for re-running recon-all after correction on WM/pial/talairach transformation etc. Would re-running fastsurfer consider any updates to the corrected file(s)? Thanks for the update!

**Comment by m-reuter** [comment]
Currently FastSurfer does not support re-running parts of the pipeline, but it is on the to-do list.

**Comment by zswgzx** [comment]
Just finished my trial for the full pipeline with just one thread. And the outputs resemble the FS ones so I guess it runs well.

One thing I noticed was that the aparc.a2009s+aseg.mgz was still missing (this was what we've been using throughout the analyses) but the corresponding label/annot & stats existed. Would it be possible to generate the actual segmentation image from the existing outputs? Thanks!

**Comment by m-reuter** [comment]
Great. The easiest would be to run recon_all from FreeSurfer with the -aparc2aseg flag. It will create those files. See here for the command it uses: https://github.com/freesurfer/freesurfer/blob/b28bb9591cb2de638f3bc560d41cf4f4e13b815a/scripts/recon-all#L4601 
So if you want only one of those files, run this one time, look at the log file and copy the exact command for the left and right hemisphere. Not sure why we don't create those. Maybe it is (or was) not part of the stream in FS 6.0 .

**Comment by zswgzx** [comment]
Just a quick question: if I pull the docker image from deepmi/fastsurfer, would the latest tag be the same as beta?

**Comment by Ouzaixin** [comment]
> ```shell
> ./run_fastsurfer.sh --t1 $t1.nii.gz --sid $id --sd $dest --py python --parallel --fsaparc
> ```

Hi, the following is my command, but I cannot get other atlas segmentation.
![image](https://github.com/user-attachments/assets/d41be10e-081e-4104-9e40-ce318a4caa55)
![image](https://github.com/user-attachments/assets/0f3b83da-3a55-4210-b89e-d5c9a7f77391)

**Comment by m-reuter** [comment]
Hi, first let us know what release version you are using. Also if native install or docker. Native installs have the disadvantage, that we don't know if you have all dependencies in the correct versions, so strange things can happen if that is not the case.

The command looks OK, but you should have more files in the mri directory. For example the wmparc.mgz that you mention above, is not here. So this cannot be the complete directory? Or did you stop early?

Also, the aparc are surface segmentations, they will be in the surf directory and corresponding stats files will be in the stats directory. 

Finally, to your other post above, the --fsaparc flag creates all the FreeSurfer surface segmentations, but it will not map them into the white matter, so you are right, the wmparc.mgz still points to the DKT atlas version. If you want also the wmparc using the aparc surface atlas, you need to run the command to create it separately. 
Should be something like this:

```
mri_surf2volseg --o $mdir/wmparc.mgz --label-wm --i $mdir/aparc+aseg.mgz --lh-annot $ldir/lh.aparc.annot 3000 --lh-cortex-mask $ldir/lh.cortex.label --lh-white $sdir/lh.white --lh-pial $sdir/lh.pial --rh-annot $ldir/rh.aparc.annot 4000 --rh-cortex-mask $ldir/rh.cortex.label --rh-white $sdir/rh.white --rh-pial $sdir/rh.pial"
 ```

 where mdir is the mri directory, ldir the label directory, sdir the surf directory.  You can also try to find this command in recon-all.log from a regular freesurfer run and copy it form there to make sure it is correct. 
 
 Generally we recommend switching to the newer DKT atlas. It is not recommended to use the aparc DK atlas for nostalgic reasons. The labels that were removed in DKT were not very stable and small, so they were removed/merged for a reason.
